keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1618 @@
+for third-party hardware vendors. Both are expected to co-exist, with vendors
+choosing the API that better suits their needs on a case-by-case basis. Because
+the new API is going to be implemented independently of the existing device
+manager, there's little risk of breaking stable APIs.
+
+The new API is inspired by the existing [volume provisioning support with CSI](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/container-storage-interface.md#provisioning-and-deleting) and uses similar
+concepts. The goal is to let users request resources with parameters that can
+be different depending on what kind of resource gets requested. Resource
+allocations can be ephemeral (specified in a Pod spec, allocated and destroyed
+together with the Pod) and persistent (lifecycle managed separately from a Pod,
+with an allocated resource used for multiple different Pods).
+
+Several core Kubernetes components must be modified (see the
+[implementation](#implementation) section for details):
+- kube-apiserver (new API)
+- kube-controller-manager (new controller)
+- kube-scheduler (new builtin scheduler plugin)
+- kubelet (new third-party kubelet plugin type)
+
+Resources as described by this KEP are managed by third-party resource drivers that communicate with central
+Kubernetes components, in particular the kube-scheduler, by updating
+objects stored in the kube-apiserver. kube-scheduler only needs to be modified
+once to support dynamic resource allocation. Then multiple drivers from
+different vendors can be installed at the same time without making further
+changes to the scheduler.
+
+Communication between the kubelet and the node part of the driver is
+handled through local Unix domain sockets and the kubelet plugin registration
+mechanism, using a new plugin type and a new gRPC interface.
+The container runtime uses the
+[Container Device Interface
+(CDI)](https://github.com/container-orchestrated-devices/container-device-interface)
+to expose the resources.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Originally, Kubernetes and its scheduler only tracked CPU and RAM as
+resources for containers. Later, support for storage and discrete,
+countable per-node extended resources was added. The kubelet device plugin
+interface then made such local resources available to containers. But
+for many newer devices, this approach and the Kubernetes API for
+requesting these custom resources is too limited. This KEP addresses
+limitations of the current approach for the following use cases:
```

Comment:
just to be clear, we are not introducing new "hooks" in pod startup to enable such use cases, all of those use cases are addressed by making it possible to pass arbitrary parameters to the driver, correct?

Reply:
Correct.